--- a/iscsiuio/src/unix/libs/bnx2x.c
+++ b/iscsiuio/src/unix/libs/bnx2x.c
@@ -120,6 +120,7 @@
  * PCI ID constants
  ******************************************************************************/
 #define PCI_VENDOR_ID_BROADCOM			0x14e4
+#define PCI_VENDOR_ID_QLOGIC			0x1077
 #define PCI_DEVICE_ID_NX2_57710			0x164e
 #define PCI_DEVICE_ID_NX2_57711			0x164f
 #define PCI_DEVICE_ID_NX2_57711E		0x1650
@@ -194,6 +195,14 @@
 	 PCI_ANY_ID, PCI_ANY_ID, brcm_57840_MF},
 	{PCI_VENDOR_ID_BROADCOM, PCI_DEVICE_ID_NX2_57840_VF,
 	 PCI_ANY_ID, PCI_ANY_ID, brcm_57840_VF},
+	{PCI_VENDOR_ID_QLOGIC, PCI_DEVICE_ID_NX2_57840_4_10,
+	 PCI_ANY_ID, PCI_ANY_ID, brcm_57840_4_10},
+	{PCI_VENDOR_ID_QLOGIC, PCI_DEVICE_ID_NX2_57840_2_20,
+	 PCI_ANY_ID, PCI_ANY_ID, brcm_57840_2_20},
+	{PCI_VENDOR_ID_QLOGIC, PCI_DEVICE_ID_NX2_57840_MF,
+	 PCI_ANY_ID, PCI_ANY_ID, brcm_57840_MF},
+	{PCI_VENDOR_ID_QLOGIC, PCI_DEVICE_ID_NX2_57840_VF,
+	 PCI_ANY_ID, PCI_ANY_ID, brcm_57840_VF},
 };
 
 static struct iro e1_iro[2] = {
