--- a/iscsiuio/src/uip/uip.c
+++ b/iscsiuio/src/uip/uip.c
@@ -220,10 +220,14 @@
 		   uip_ip4addr_t *netmask,
 		   uip_ip4addr_t *default_route, uint8_t *mac_addr)
 {
-	uip_sethostaddr4(ustack, ip);
-	uip_setnetmask4(ustack, netmask);
-	uip_setdraddr4(ustack, default_route);
-	uip_setethernetmac(ustack, mac_addr);
+	if (ip)
+		uip_sethostaddr4(ustack, ip);
+	if (netmask)
+		uip_setnetmask4(ustack, netmask);
+	if (default_route)
+		uip_setdraddr4(ustack, default_route);
+	if (mac_addr)
+		uip_setethernetmac(ustack, mac_addr);
 }
 
 #if !UIP_ARCH_ADD32
--- a/iscsiuio/src/unix/nic.c
+++ b/iscsiuio/src/unix/nic.c
@@ -470,9 +470,11 @@
 	pthread_mutex_lock(&nic->nic_mutex);
 
 	/*  Check if the file node exists before closing */
-	rc = stat(nic->uio_device_name, &file_stat);
-	if ((rc == 0) && (nic->ops))
-		nic->ops->close(nic, 0);
+	if (nic->uio_device_name) {
+		rc = stat(nic->uio_device_name, &file_stat);
+		if ((rc == 0) && (nic->ops))
+			nic->ops->close(nic, 0);
+	}
 	pthread_mutex_unlock(&nic->nic_mutex);
 
 	nic->state = NIC_EXIT;
@@ -1268,17 +1270,13 @@
 			 nic->log_name, inet_ntoa(addr));
 
 		set_uip_stack(&nic_iface->ustack,
-			      &nic_iface->ustack.hostaddr,
-			      &nic_iface->ustack.netmask,
-			      &nic_iface->ustack.default_route_addr,
+			      NULL, NULL, NULL,
 			      nic_iface->mac_addr);
 		break;
 
 	case IPV4_CONFIG_DHCP:
 		set_uip_stack(&nic_iface->ustack,
-			      &nic_iface->ustack.hostaddr,
-			      &nic_iface->ustack.netmask,
-			      &nic_iface->ustack.default_route_addr,
+			      NULL, NULL, NULL,
 			      nic_iface->mac_addr);
 		if (dhcpc_init(nic, &nic_iface->ustack,
 			       nic_iface->mac_addr, ETH_ALEN)) {
