Index: open-iscsi/usr/Makefile
===================================================================
--- open-iscsi.orig/usr/Makefile	2011-08-22 20:25:31.643344958 +0530
+++ open-iscsi/usr/Makefile	2011-08-22 20:26:24.120647120 +0530
@@ -31,7 +31,7 @@
 OPTFLAGS ?= -O2 -g
 WARNFLAGS ?= -Wall -Wstrict-prototypes
 CFLAGS += $(OPTFLAGS) $(WARNFLAGS) -I../include -I. -I../utils/open-isns \
-					-D$(OSNAME) $(IPC_CFLAGS)
+					-D$(OSNAME) $(IPC_CFLAGS) -DISNS_ENABLE
 PROGRAMS = iscsid iscsiadm iscsistart
 
 # libc compat files
Index: open-iscsi/usr/discovery.c
===================================================================
--- open-iscsi.orig/usr/discovery.c	2011-08-22 20:25:36.695470434 +0530
+++ open-iscsi/usr/discovery.c	2011-08-22 20:26:24.120647120 +0530
@@ -44,9 +44,11 @@
 #include "iscsid_req.h"
 #include "iscsi_util.h"
 /* libisns includes */
+#ifdef ISNS_ENABLE
 #include "isns.h"
 #include "paths.h"
 #include "message.h"
+#endif
 
 #ifdef SLP_ENABLE
 #include "iscsi-slp-discovery.h"
@@ -93,6 +95,7 @@
 	return 0;
 }
 
+#ifdef ISNS_ENABLE
 void discovery_isns_free_servername(void)
 {
 	if (isns_config.ic_server_name)
@@ -366,6 +369,7 @@
 	discovery_isns_free_servername();
 	return rc;
 }
+#endif
 
 int discovery_fw(void *data, struct iface_rec *iface,
 		 struct list_head *rec_list)
